### PR TITLE
Support batched `pinv`

### LINKED ISSUE #3062: cupy.linalg.{svd, pinv} should support broadcasting
Problem
-----------
The NumPy implementation of [pinv](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.pinv.html) supports broadcasting, but the CuPy implementation does not.

Conditions
--------------
```
CuPy Version          : 7.0.0
CUDA Root             : /local/cuda-10.0
CUDA Build Version    : 10010
CUDA Driver Version   : 10020
CUDA Runtime Version  : 10010
cuBLAS Version        : 10201
cuFFT Version         : 10101
cuRAND Version        : 10101
cuSOLVER Version      : (10, 2, 0)
cuSPARSE Version      : 10300
NVRTC Version         : (10, 1)
cuDNN Build Version   : 7604
cuDNN Version         : 7605
NCCL Build Version    : 2406
NCCL Runtime Version  : 2506
```
Reproduce
--------------
```python
import numpy as np
import cupy as cp

A = np.random.rand(3, 4, 5)
xA = cp.array(A)
np.linalg.pinv(A)
cp.linalg.pinv(xA)
```
```
Traceback (most recent call last):
  File "pinv.py", line 7, in <module>
    cp.linalg.pinv(xA)
  File "/home/beams/DCHING/miniconda3/envs/tike/lib/python3.7/site-packages/cupy/linalg/solve.py", line 421, in pinv
    u, s, vt = decomposition.svd(a.conj(), full_matrices=False)
  File "/home/beams/DCHING/miniconda3/envs/tike/lib/python3.7/site-packages/cupy/linalg/decomposition.py", line 245, in svd
    util._assert_rank2(a)
  File "/home/beams/DCHING/miniconda3/envs/tike/lib/python3.7/site-packages/cupy/linalg/util.py", line 20, in _assert_rank2
    'two-dimensional'.format(a.ndim))
numpy.linalg.LinAlgError: 3-dimensional array given. Array must be two-dimensional
```
Solution
-----------
Change CuPy implementation to allow broadcasting on N+2 dimensional arrays. This may also require updating `cupy.linalg.svd` because `pinv` relies on `svd` and NumPy allows broadcasting for that function too.

### PR DESCRIPTION
**Pull request by leofang**
~~***Blocked by #4684*** as batched SVD is required but currently emitting random errors in the CI.~~

Close #3062.

- For supporting batched `pinv`: the original implementation already followed NumPy's, so really it's just adding `newaxis` here and there to broadcast. 
- This PR also fixes an incompatibility when the input array is of zero size

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

**Review by toslunar — CHANGES_REQUESTED** [review]

File: tests/cupy_tests/linalg_tests/test_solve.py [hunk 1]
```diff
@@ -177,11 +176,22 @@ def test_pinv(self):
...
+    def test_pinv_batched_vector_rcond(self):
+        self.check_x((2, 3, 4), rcond=[0.5, 0.5])
+        self.check_x((2, 3, 4, 5),
+                     rcond=[[0.2, 0.2, 0.2],
+                            [0.2, 0.2, 0.2]])
+
+    def test_pinv_size_0(self):
+        self.check_x((3, 0), rcond=1e-15)
+        self.check_x((0, 3), rcond=1e-15)
+        self.check_x((0, 0), rcond=1e-15)
```
[the remark this answers is not in the document]
Comment (leofang) [reply tests/cupy_tests/linalg_tests/test_solve.py:192]: Thanks, added it!

File: tests/cupy_tests/linalg_tests/test_solve.py [hunk 2]
```diff
@@ -177,11 +176,22 @@ def test_pinv(self):
...
-        self.check_shape((4, 3, 2, 1))
+    def test_pinv_batched(self):
+        self.check_x((2, 3, 4), rcond=1e-15)
+        self.check_x((2, 3, 4, 5), rcond=1e-15)
+
+    def test_pinv_batched_vector_rcond(self):
+        self.check_x((2, 3, 4), rcond=[0.5, 0.5])
+        self.check_x((2, 3, 4, 5),
+                     rcond=[[0.2, 0.2, 0.2],
+                            [0.2, 0.2, 0.2]])
```
[the remark this answers is not in the document]
Comment (leofang) [reply tests/cupy_tests/linalg_tests/test_solve.py]: Updated in da610ee.

**Comment by leofang** [comment]
Thanks, @toslunar! Your suggestions make it much cleaner 👍

**Comment by leofang** [comment]
Jenkins, test this please

**Review by toslunar — APPROVED** [review]

**Comment by leofang** [comment]
Thanks, @toslunar!

**Pull request merged**